### PR TITLE
Add extended test coverage for Qwen2-VL vision transformer

### LINKED ISSUE #3556: [Feature]: Expand test coverage for Qwen2VL vision-language model
### 🚀 Feature Description

Expand test coverage for the `Qwen2VL` vision-language model, which currently has **minimal testing** with only 33 test lines for 205 lines of implementation (0.16 ratio), consisting of smoke tests.

### 📂 Feature Category

VLM/VLA Models (Vision Language Models/Agents) - Priority

### 💡 Motivation

**Current situation:**
- `Qwen2VL` has only **33 test lines** (205 lines of implementation)
- Current tests consist of only a basic smoke test
- **Missing critical tests:**
  - No gradient checks (gradcheck)
  - No component-level tests
  - No torch.compile/dynamo tests
  - No integration tests with actual vision-language tasks
  - No pretrained weight loading verification (if applicable)
  - No batch consistency tests
  - No exception handling tests

**Why expanded testing is needed:**
- Verify vision-language model functionality beyond basic instantiation
- Ensure compatibility with PyTorch optimization features (torch.compile)
- Test gradient flow through vision and language components
- Prevent regressions in model architecture changes
- Provide comprehensive usage examples
- Match testing standards of other VLMs (e.g., KimiVL: 0.64 ratio)

### 💭 Proposed Solution

**Migrate to `BaseTester` pattern:**
- ✅ Smoke tests: Multiple configurations, batch sizes, input formats
- ✅ Exception tests: Invalid inputs, edge cases, error handling
- ✅ Cardinality tests: Output shape verification for various inputs
- ✅ Component tests: Individual module testing (vision encoder, language decoder, etc.)
- ✅ Feature tests: Vision-language alignment, attention mechanisms
- ✅ Gradient checks: Verify backpropagation correctness using `gradcheck`
- ✅ Torch.compile compatibility: Test with `torch_optimizer` fixture
- ✅ Integration tests: End-to-end vision-language tasks
- ✅ Batch consistency: Verify batch vs. individual processing
- ✅ Pretrained weights: Verify weight loading if pretrained models available

### Additional Documentation

**Jupyter notebook demonstrating:**
1. Model instantiation and configuration
2. Vision-language inference examples
3. Image-text alignment capabilities
4. Comparison with official Qwen2VL implementation
5. Performance benchmarks
6. Integration with Hugging Face models (if applicable)

### 🔄 Alternatives Considered

_No response_

### 🎯 Use Cases

**For developers:**
- Ensure vision-language model works correctly after changes
- Test gradient flow through model components
- Verify compiler optimization compatibility
- Understand model architecture through tests

**For users:**
- Learn proper usage patterns for vision-language tasks
- See practical examples with images and text
- Understand model capabilities and limitations
- Get copy-paste examples for inference

**For maintainers:**
- Maintain code quality standards across VLMs
- Catch regressions before deployment
- Ensure consistency with other VLM implementations

### 📝 Additional Context

**Gap identified:** Qwen2VL has the lowest test coverage (0.16) among VLMs in the repository.

### 🤝 Contribution Intent

- [ ] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

### PR DESCRIPTION
**Pull request by Ayushi6002**
## 🚀 Summary

This PR extends test coverage for the Qwen2-VL vision transformer module.
Fixes #3556

## ✅ Changes

- Added smoke test to validate forward pass and output shape
- Added gradient test to verify backward propagation
- Added batch consistency test (single vs batch inference)
- Added patch merger shape validation test
- Avoided shadowing Python built-ins (`input` → `x`)
- Reduced model size in tests for stability and faster execution

## ⚠️ Notes

- `torch.compile` test is skipped due to instability on Windows with Python 3.13 (causes access violation)

## 📌 Result

All tests pass successfully on CPU.

## 🔍 Motivation

Improves reliability and correctness of the Qwen2-VL implementation by ensuring:
- Proper gradient flow
- Stable behavior across batch sizes
- Correct output dimensions

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [x] 🟢 **No AI used.**
- [ ] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [ ] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

### CODE REVIEW

File: tests/models/qwen25/test_qwen2_vl.py [hunk 1]
```diff
@@ -18,16 +18,53 @@
...
+    Qwen2VLVisionTransformer,
+)
 
 
 class TestQwen2VL:
     @pytest.mark.parametrize("batch_size", [1, 2])
     def test_smoke(self, batch_size, device, dtype):
-        model = Qwen2VLVisionTransformer().to(device=device, dtype=dtype)
-        input = torch.randn(batch_size, 3, 224, 224, device=device, dtype=dtype)
+        # 🔥 smaller model → stable
```
Comment (Meggison) [inline tests/models/qwen25/test_qwen2_vl.py]: Could you take out the fire emoji, doesn't fit with formatting

File: tests/models/qwen25/test_qwen2_vl.py [hunk 2]
```diff
@@ -18,16 +18,53 @@
...
+
+        x = torch.randn(1, 3, 224, 224, device=device, requires_grad=True)
+        output = model(x)
+
+        loss = output.mean()
+        loss.backward()
+
+        assert x.grad is not None
+
+    def test_batch_consistency(self, device):
```
Comment (Meggison) [inline tests/models/qwen25/test_qwen2_vl.py:58]: Could you add docstrings for explainability,

**Review by Meggison — CHANGES_REQUESTED** [review]

**Comment by Meggison** [comment]
Please rebase as well @Ayushi6002

**Comment by Ayushi6002** [comment]
@Meggison and @sidd-27 all requested changes have been addressed.
Rebased on the latest `main` and all CI checks (tests, tutorials, pre-commit) are now passing.

Please let me know if anything else is needed — otherwise ready for merge.

File: tests/models/qwen25/test_qwen2_vl.py [hunk 3]
```diff
@@ -18,16 +18,54 @@
...
+        merger = Qwen2VLPatchMerger(dim=64).to(device)
+
+        x = torch.randn(1, 3, 224, 224, device=device)
+        output = merger(x)
+
+        assert output.shape == (1, 256, 64)
+
+    @pytest.mark.skip(reason="torch.compile unstable on Windows/Python 3.13")
+    def test_compile(self, device):
+        pass
```
Comment (sidd-27) [inline tests/models/qwen25/test_qwen2_vl.py]: why has this been added?

**Review by sidd-27 — CHANGES_REQUESTED** [review]

File: tests/models/qwen25/test_qwen2_vl.py (hunk 3 above)
Comment (Ayushi6002) [reply tests/models/qwen25/test_qwen2_vl.py]: This test was added while experimenting with torch.compile, but since it doesn’t add meaningful coverage and was skipped, I’ve removed it to keep the test suite clean.

Thanks for pointing it out!

**Review by ducha-aiki — APPROVED** [review]
